File: comp-arch-2/src/lower_triangular_matrix.cpp
//------------------------------------------------------------------------------
// lower_triangular_matrix.cpp - содержит функции нижнетреугольной матрицы
//------------------------------------------------------------------------------

#include "lower_triangular_matrix.h"

//--------------------------------------------
// Ввод элементов нижнетреугольной матрицы из потока
void LTMatrix::In(std::ifstream &ifst) {
    ifst >> dimensionality;
    int size = dimensionality;
    int col = 0;
    int row = 0;
    for (int i = 0; i < size * size; ++i) {
        col = i % size;
        row = i / size;

        if (col <= row) {
            ifst >> elements[i];
        } else {
            elements[i] = 0;
        }

        ++col;
    }
}

//------------------------------------------------------
// Случайный ввод элементов нижнетреугольной матрицы из потока
void LTMatrix::InRnd() {
    dimensionality = Random();
    int size = dimensionality;
    int col = 0;
    int row = 0;
    for (int i = 0; i < size * size; ++i) {
        col = i % size;
        row = i / size;

        if (col <= row) {
            elements[i] = Random();
        } else {
            elements[i] = 0;
        }

        ++col;
    }
}

//---------------------------------------------
// Вывод элементов нижнетреугольной матрицы из потока
void LTMatrix::Out(std::ofstream &ofst) {
    int size = dimensionality;
    int col = 0;
    ofst << "Lower triangular matrix:\n"
         << "Dimensionality: " << dimensionality << "\n"
         << "Arithmetic mean: " << ArithmeticMean() << "\n"
         << "Elements: \n";
    for (int i = 0; i < size * size; ++i) {
        col = i % size;

        ofst << elements[i] << " ";

        ++col;

        if (col == size) {
            ofst << "\n";
        }
    }
}

//--------------------------------------------
// Вычисление среднего арифметического нижнетреугольной матрицы
double LTMatrix::ArithmeticMean() {
    int size = dimensionality;
    int count = size * size;
    double sum = 0;
    int col = 0;

    for (int i = 0; i < size * size; ++i) {
        col = i % size;

        sum += elements[i];

        ++col;
    }

    return sum / count;
}

File: comp-arch-3/src/lower_triangular_matrix.py
from typing import IO
from matrix import Matrix
from random_generator import get_rand_int


class LTMatrix(Matrix):
    def __init__(self, dimensionality: int):
        super().__init__(dimensionality)
        self.matrix_data.elements = []

    def input(self, read_file: IO) -> IO:
        size: int = self.matrix_data.dimensionality

        all_data = list(map(float, read_file.readline().split()))
        for i in range(1, size + 1):
            for _ in range(i):
                self.matrix_data.elements.append(all_data.pop())
            self.matrix_data.elements.extend([0] * (size - i))

        return read_file

    def input_random(self) -> None:
        size: int = self.matrix_data.dimensionality

        self.matrix_data.elements = [get_rand_int() for _ in range(size ** 2)]

    def output(self, write_file: IO) -> None:
        size: int = self.matrix_data.dimensionality
        col: int = 0

        write_file.write(
            f'Lower triangular matrix:\n'
            f'Dimensionality: {size}\n'
            f'Arithmetic mean: {self.arithmetic_mean()}\n'
            f'Elements: \n'
        )

        for i in range(size ** 2):
            col = i % size

            write_file.write(
                f'{self.matrix_data.elements[i]}'
                + (' ' if col != size - 1 else '\n')
            )

            col += 1

    def arithmetic_mean(self) -> float:
        size: int = self.matrix_data.dimensionality
        count: int = size ** 2
        elements_sum: int = 0
        col: int = 0

        for i in range(size ** 2):
            col = i % size

            elements_sum += self.matrix_data.elements[i]

            col += 1

        return elements_sum / count
